### PR TITLE
🎉 New Destination: Datastax Astra

### PR DESCRIPTION
**Pull request by anomnaco**
New destination for vector database Datastax Astra via the new JSON API. Requires python cdk to include embedder and other vector db utilities included in the cdk.

### CODE REVIEW

**Review by flash1293 — COMMENTED** [review]
Hey, this PR looks great already! I left some higher level comments but there is nothing big in there. Let me know once I can take another look

File: airbyte-integrations/connectors/destination-astra/setup.py
Comment (anomnaco) [reply airbyte-integrations/connectors/destination-astra/setup.py]: Fixed

File: airbyte-integrations/connectors/destination-astra/destination_astra/indexer.py
Comment (anomnaco) [reply airbyte-integrations/connectors/destination-astra/destination_astra/indexer.py:42]: Fixed

File: airbyte-integrations/connectors/destination-astra/destination_astra/astra_client.py
Comment (anomnaco) [reply airbyte-integrations/connectors/destination-astra/destination_astra/astra_client.py]: Done

File: airbyte-integrations/connectors/destination-astra/integration_tests/integration_test.py
Comment (anomnaco) [reply airbyte-integrations/connectors/destination-astra/integration_tests/integration_test.py:31]: Astra does have a free tier. Confirmed with Datastax that it gives access to vector search and the JSON API so this should be doable.

File: docs/integrations/destinations/astra.md
Comment (anomnaco) [reply docs/integrations/destinations/astra.md:1]: Updated

File: airbyte-integrations/connectors/destination-astra/destination_astra/destination.py
Comment (flash1293) [inline airbyte-integrations/connectors/destination-astra/destination_astra/destination.py:43]: This should be in the indexers `check` method instead. Exceptions thrown from here won't resolve in a nice error message because they are not handled (the indexer `check` method already has a try-catch clause to print a nice error message.

File: docs/integrations/destinations/astra.md
Comment (flash1293) [inline docs/integrations/destinations/astra.md:23]: Do we need this section with the new logic of the connector auto-creating the collection? If there's a good reason to still do it manually in some cases we can leave it in with a note that the connector can also take care of it, wdyt?

File: docs/integrations/destinations/astra.md
Comment (flash1293) [inline docs/integrations/destinations/astra.md:43]: The connector supports all sync modes (the vector DB CDK takes care of this if you implement `index` and `delete` on the indexer properly which you did. You can use the same table as here: https://github.com/airbytehq/airbyte/blob/6dff143e1d68038440b4db88784038e9b3d1bd3a/docs/integrations/destinations/chroma.md#L10

File: airbyte-integrations/connectors/destination-astra/destination_astra/spec.json
Comment (flash1293) [inline airbyte-integrations/connectors/destination-astra/destination_astra/spec.json:1]: This file should be in the integration_tests folder (don't forget to adjust the reference from `acceptance-test-config.yml`)

**Review by flash1293 — COMMENTED** [review]
Played around with it and it works pretty well, great job!

Had some smaller comments left.

Also, I tried to run the unit tests and they are failing for me, could you take a look at that?

File: airbyte-integrations/connectors/destination-astra/integration_tests/integration_test.py
Comment (alexleventer) [reply airbyte-integrations/connectors/destination-astra/integration_tests/integration_test.py:31]: Hey @flash1293 - let me know if you need anymore help with this. I think we resolved this by turning off auto-hibernation for your org.

File: airbyte-integrations/connectors/destination-astra/destination_astra/config.py
Comment (anomnaco) [reply airbyte-integrations/connectors/destination-astra/destination_astra/config.py]: Fixed

File: airbyte-integrations/connectors/destination-astra/destination_astra/config.py
Comment (anomnaco) [reply airbyte-integrations/connectors/destination-astra/destination_astra/config.py]: Fixed

File: airbyte-integrations/connectors/destination-astra/destination_astra/destination.py
Comment (anomnaco) [reply airbyte-integrations/connectors/destination-astra/destination_astra/destination.py]: Moved

File: docs/integrations/destinations/astra.md
Comment (anomnaco) [reply docs/integrations/destinations/astra.md]: Removed

File: docs/integrations/destinations/astra.md
Comment (anomnaco) [reply docs/integrations/destinations/astra.md]: Fixed

**Comment by hc33brackles** [comment]
Thanks for this feedback @flash1293 -- we are close to resolving all of these issues and updating the branch with the correction. I'll check back in on Monday - there's still a final issue we are working on resolving.

**Comment by anomnaco** [comment]
Moved create collection back to destination.py because it was causing issues with integration and unit tests.

File: airbyte-integrations/connectors/destination-astra/destination_astra/spec.json
Comment (anomnaco) [reply airbyte-integrations/connectors/destination-astra/destination_astra/spec.json:1]: Moved

**Comment by hc33brackles** [comment]
@flash1293 - we are done with your most recent comments as well. Please let me know what the next steps are!

File: airbyte-integrations/connectors/destination-astra/destination_astra/destination.py
Comment (flash1293) [inline airbyte-integrations/connectors/destination-astra/destination_astra/destination.py:42]: On https://github.com/airbytehq/airbyte/pull/34058/files#r1459806142 you wrote you moved the `create_collection` call, but it seems like some merge foo brought it back (it's also not in the indexer). Could you double check please?

File: airbyte-integrations/connectors/destination-astra/destination_astra/config.py
Comment (flash1293) [inline airbyte-integrations/connectors/destination-astra/destination_astra/config.py:19]: Could we keep the endpoint non-secret? This is causing an issue in the UI when editing an existing configuration:
Screenshot 2024-01-23 at 18 07 47

I think nobody ran into this problem yet. If something is marked as secret, two things happen:
* it's not shown in the UI in cleartext
* It is stored in a separate secret manager, not in the regular Airbyte DB with the rest of the configuration

I might miss something, but it seems like the endpoint isn't sensitive in this sense and can be stored with the rest of the configuration (the app token definitely is and should be declared a secret)

**Review by flash1293 — COMMENTED** [review]
Hi @hc33brackles and @anomnaco , apologies, I was off yesterday and couldn't give it another look.

This is starting to look very good! I updated a few small things:
* Apply formatting
* Add integration test secret to our CI
* Update spec for integration test
* Prepare the icon in our platform repo so it will show properly

CI is currently running here: https://github.com/airbytehq/airbyte/pull/34442

Let's see whether it works. I need to do this on a separate PR for permission reasons. Could you give me write access to the `destination-astra` on the `Anant` fork so I can push the same things over there (or do it yourself)? This will make it easier to wrap this work up.

Left two smaller comments in the PR

File: airbyte-integrations/connectors/destination-astra/destination_astra/destination.py
Comment (anomnaco) [reply airbyte-integrations/connectors/destination-astra/destination_astra/destination.py:42]: Its on the next line here

File: airbyte-integrations/connectors/destination-astra/destination_astra/config.py
Comment (anomnaco) [reply airbyte-integrations/connectors/destination-astra/destination_astra/config.py]: Done

File: airbyte-integrations/connectors/destination-astra/destination_astra/destination.py
Comment (flash1293) [reply airbyte-integrations/connectors/destination-astra/destination_astra/destination.py:42]: @anomnaco yes, it shouldn’t be here in the destination class, it should be in the indexer class check method so it’s properly wrapped in a try/except block and a meaningful error message is produced on check

Comment (anomnaco) [reply airbyte-integrations/connectors/destination-astra/destination_astra/destination.py:42]: That caused issues with the unit and integration tests

Comment (hc33brackles) [reply airbyte-integrations/connectors/destination-astra/destination_astra/destination.py:42]: Hi @flash1293 - I've been chatting with Obi about this offline. He was testing with this in the indexer class check method and got failures with the unit tests with it there. Do you need us to look deeper into getting this to work in the way that you suggested or is it acceptable here?

Comment (hc33brackles) [reply airbyte-integrations/connectors/destination-astra/destination_astra/destination.py:42]: We are working on resolving this now.

Comment (flash1293) [reply airbyte-integrations/connectors/destination-astra/destination_astra/destination.py:42]: The important point to me is that the user gets an actionable error message on check as part of a well-formed status message and the connector doesn't crash as this won't produce a nice output in the UI.

**Comment by flash1293** [comment]
@anomnaco @hc33brackles About the problem that's still left - maybe the create_collection call needs to be mocked out in the unit test (plus validation it was called as expected)? https://github.com/airbytehq/airbyte/pull/34058/files#diff-ad0e480b627023296a1ae9ff8e684e73cf34f8337430d07d549b65ae583158acR157

**Review by flash1293 — APPROVED** [review]
I had to fix the unit tests as the create_collection call needs to be mocked out now, but tested again and everything seems to work well, LGTM.

We have a CI issue right now, as soon as that's resolved I'm going to merge

**Pull request merged**
